### PR TITLE
OpenAI implementation for Google ADK

### PR DESCRIPTION
**Pull request by secprog**
### Summary

Adds OpenAI LLM integration for the ADK framework using OpenAI's Responses API (2025). Supports file uploads, multimodal content, metadata extraction, and Google GenAI config feature conversions.

#### 1. Responses API Integration
- Uses `client.responses.create()` for all requests
- Supports streaming and non-streaming responses
- Handles Responses API format with `output[]` containing message/function/tool_result items
- Uses `input` parameter for messages and `instructions` for system prompts
- Parses streaming events: `response.created`, `response.content_part.added`, `response.function_call.*`, `response.done`

#### 2. Files API Support
- Document uploads (PDFs, Word docs) via Files API
- Automatic fallback to base64 encoding if upload fails
- `use_files_api: bool = True` attribute (default enabled)
- Temporary file management with automatic cleanup
- Returns `{"type": "file", "file_id": "..."}` format for Responses API

#### 3. Multimodal Content Support
- Image handling: data URIs (`data:image/png;base64,...`) and external URLs
- Automatic content fetching from URLs using `aiohttp` or `httpx`
- MIME type detection from URL extension or data URI header
- Conversion to Google GenAI `types.Part(inline_data=types.Blob(...))`
- Supports PNG, JPEG, GIF, WebP formats

#### 4. Response Metadata Extraction
- Citation metadata from `response.citations`, `response.sources`, or `response.metadata.citations`
- Safety/moderation metadata from `response.safety`, `response.moderation`, or `response.content_filter`
- Cache metadata from `response.cache` or `response.metadata.cache`
- Tool result handling with citation extraction
- Metadata extraction in both streaming and non-streaming responses

#### 5. Google GenAI Config Feature Conversions
- `response_logprobs`: Converts boolean to integer `logprobs` (defaults to 5 if not set)
- `grounding_config.web_search`: Automatically adds `web_search` tool to tools list
- `caching_config`: Logs debug messages (cache behavior is automatic in Responses API)
- `safety_settings`: Logs debug messages (moderation is built-in and not configurable)

#### 6. Structured Outputs Support
- JSON Mode: Maps `response_mime_type="application/json"` to `{"type": "json_object"}`
- JSON Schema Mode: Converts `response_schema` to OpenAI's `json_schema` format
- Pydantic Model Support: Automatically extracts schemas from Pydantic v1/v2 models
- Schema Name Extraction: From class names, title fields, or defaults to "ResponseSchema"
- OpenAI API Compliance: Wraps schemas with required name field

#### 7. Tool/Function Calling
- Tool Declaration Conversion: Converts Google GenAI `FunctionDeclaration` to OpenAI tool format
- Type System Mapping: Handles Google GenAI type enums and converts to OpenAI string types
- Tool Choice Configuration: Maps `FunctionCallingConfigMode` to OpenAI `tool_choice`
- Parallel Tool Calls: Automatically sets `parallel_tool_calls=True`
- Message Sequencing: Validates and handles tool call/response message sequences

#### 8. Parameter Mapping (Google GenAI → OpenAI Responses API)

| Google GenAI Field | OpenAI Parameter | Mapping Logic |
|-------------------|------------------|---------------|
| `max_output_tokens` | `max_output_tokens` | Direct mapping |
| `stop_sequences` | `stop` | Direct list mapping |
| `candidate_count` | `n` | Direct mapping |
| `tool_config.function_calling_config.mode` | `tool_choice` | Enum to string mapping |
| `response_schema` | `response_format.json_schema` | Schema wrapping with name |
| `response_mime_type` | `response_format.type` | MIME type to format mapping |
| `response_logprobs` | `logprobs` | Boolean to integer conversion |
| `grounding_config.web_search` | `tools` (web_search) | Adds web_search tool |
| `system_instruction` | `instructions` | Direct mapping |
| `temperature`, `top_p`, `frequency_penalty`, `presence_penalty`, `seed`, `logprobs` | Direct mappings | Direct mapping |

### Implementation Details

**Strict Google GenAI Type Usage**
- All parameters sourced from `GenerateContentConfig`
- Uses `tool_config.function_calling_config.mode` instead of direct `tool_choice`
- Uses `stop_sequences` instead of direct `stop`
- Type-safe parameter extraction with `getattr()` for optional fields

**Error Handling & Validation**
- Model name validation against supported patterns
- Tool call/response sequence validation
- Warning logs for missing tool responses
- Graceful handling of invalid JSON in tool call arguments
- Debug logging for message sequences and parameter mapping
- Error extraction from `response.error` with code and message

**Azure OpenAI Support**
- Automatic detection via `AZURE_OPENAI_ENDPOINT` or Azure-looking URLs
- Uses `AsyncAzureOpenAI` client for Azure endpoints
- Supports Azure-specific API version configuration
- Handles Azure deployment names as model identifiers
- Requires `openai>=2.7.2` for Azure async support

**Streaming Response Handling**
- Accumulates text content across `response.content_part.added` events
- Accumulates tool calls across `response.function_call.*` events
- Extracts final metadata from `response.done` event
- Yields partial responses during streaming
- Final response includes all accumulated content and metadata

### Testing

- All 68 tests pass
- Coverage includes:
  - Responses API format parsing
  - Files API upload and fallback
  - Multimodal content conversion
  - Metadata extraction
  - Config feature conversions
  - Streaming and non-streaming responses
  - Tool/function calling
  - Structured outputs
  - Parameter mapping and type conversions

### Configuration

**Environment Variables**
- `OPENAI_API_KEY` - OpenAI API key
- `AZURE_OPENAI_API_KEY` - Azure OpenAI API key
- `OPENAI_BASE_URL` - Custom OpenAI base URL
- `AZURE_OPENAI_ENDPOINT` - Azure endpoint URL
- `AZURE_OPENAI_API_VERSION` - Azure API version (default: `"2024-02-15-preview"`)

**GenerateContentConfig Usage**

```python
config = types.GenerateContentConfig(
    max_output_tokens=200,
    temperature=0.7,
    stop_sequences=["\n", "END"],
    tool_config=types.ToolConfig(
        function_calling_config=types.FunctionCallingConfig(
            mode=types.FunctionCallingConfigMode.AUTO
        )
    ),
    response_schema=MyPydanticModel,  # or dict
    response_logprobs=True,  # Enables logprobs (defaults to 5)
    # Optional: grounding_config, caching_config, safety_settings
)
```

**OpenAI Class Attributes**
- `model: str = "gpt-4o"` - Default model name
- `use_files_api: bool = True` - Enable Files API for document uploads

### Files Changed

- `src/google/adk_community/models/openai_llm.py` (1,518 lines)
  - Implementation with Responses API integration
  - Files API support for document uploads
  - Multimodal content conversion
  - Metadata extraction (citations, safety, cache)
  - Config feature conversions
  - Strict Google GenAI type usage
  - Error handling

- `tests/unittests/models/test_model_openai.py` (1,843 lines)
  - 68 unit tests
  - Coverage for all features and edge cases
  - Tests for Responses API format
  - Tests for parameter mapping and type conversions
  - Tests for Files API and multimodal support

### Benefits

- Modern API: Uses Responses API with enhanced features
- File Support: Native Files API integration for document processing
- Multimodal: Support for images and media content
- Metadata Rich: Extracts citations, safety, and cache metadata
- Type Safety: Strict Google GenAI type usage
- API Compliance: Proper handling of Responses API requirements
- Maintainability: Clear separation of concerns with conversion functions
- Extensibility: Easy to add support for new OpenAI features
- Test Coverage: Comprehensive test suite
- Feature Conversion: Maps Google GenAI features to OpenAI equivalents where possible

All tests pass (68/68). The implementation is production-ready and fully compatible with the ADK framework's Google GenAI type system, using OpenAI's Responses API.

### CODE REVIEW

File: src/google/adk_community/models/openai_llm.py [hunk 1]
```diff
@@ -0,0 +1,1268 @@
...
+                            logger.debug(
+                                f"Response MIME type {response_mime_type} "
+                                "not directly mapped to OpenAI response_format"
+                            )
+            except Exception as e:
+                # If there's any issue accessing config attributes, log and continue
+                logger.debug(f"Error checking structured output config: {e}")
+                # Re-raise to see what's happening in tests
+                if os.getenv("DEBUG_STRUCTURED_OUTPUT"):
+                    raise
```
[the remark this answers is not in the document]
Comment (secprog) [reply src/google/adk_community/models/openai_llm.py]: fixed

File: tests/unittests/models/test_model_openai.py [hunk 2]
```diff
@@ -0,0 +1,1818 @@
...
+    @patch("openai.AsyncAzureOpenAI")
+    def test_get_openai_client_azure_fallback(self, mock_azure_openai_class):
+        """Test that ImportError is raised when AsyncAzureOpenAI not available."""
+        mock_azure_openai_class.side_effect = AttributeError(
+            "No AsyncAzureOpenAI"
+        )
+        openai_client = OpenAI()
+        with pytest.raises(ImportError) as exc_info:
+            openai_client._get_openai_client()
+        assert "openai>=2.7.2" in str(exc_info.value)
```
[the remark this answers is not in the document]
Comment (secprog) [reply tests/unittests/models/test_model_openai.py]: @gemini-code-assist you are wrong if you check pypi you will see:

Release history [Release notifications](https://pypi.org/help/#project-release-notifications) | RSS feed
This version

2.8.1

Nov 17, 2025

2.8.0

Nov 13, 2025

2.7.2

Nov 10, 2025

2.7.1

Nov 4, 2025

2.7.0

Nov 3, 2025

2.6.1

Oct 24, 2025

2.6.0

Oct 20, 2025

2.5.0

Oct 17, 2025

2.4.0

Oct 16, 2025

2.3.0

Oct 10, 2025

2.2.0

Oct 6, 2025

2.1.0

Oct 2, 2025

2.0.1

Oct 1, 2025

2.0.0

Sep 30, 2025

1.109.1

Sep 24, 2025

1.109.0

Sep 23, 2025

1.108.2

Sep 23, 2025

1.108.1

Sep 19, 2025

1.108.0

Sep 17, 2025

File: src/google/adk_community/models/openai_llm.py [hunk 3]
```diff
@@ -0,0 +1,1517 @@
...
+                                            "arguments": getattr(func_call, "arguments", ""),
+                                        },
+                                    }
+                        
+                        # Handle function arguments delta events
+                        elif event_type == "response.function_call.arguments_delta":
+                            if hasattr(event, "delta") and hasattr(event, "call_id"):
+                                call_id = event.call_id
+                                if call_id in accumulated_tool_calls:
+                                    delta_text = getattr(event.delta, "text", "") or getattr(event.delta, "", "")
```
[the remark this answers is not in the document]
Comment (secprog) [reply src/google/adk_community/models/openai_llm.py]: fixed

File: pyproject.toml [hunk 4]
```diff
@@ -29,6 +29,7 @@ dependencies = [
   "google-adk", # Google ADK
   "httpx>=0.27.0, <1.0.0", # For OpenMemory service
   "redis>=5.0.0, <6.0.0", # Redis for session storage
+  "openai>=2.7.2, <3.0.0", # OpenAI for LLM integration
```
[the remark this answers is not in the document]
Comment (secprog) [reply pyproject.toml]: @gemini-code-assist
An extensive discussion occurred between secprog and gemini-code-assist[bot] regarding the correct openai package versioning. gemini-code-assist[bot] initially incorrectly asserted that 2.x.x versions of the openai library did not exist. secprog successfully provided evidence from PyPI (releases up to 2.8.1), leading the bot to retract its suggestion and confirm the openai>=2.7.2, <3.0.0 dependency was correct for both pyproject.toml and the ImportError message.

File: src/google/adk_community/models/openai_llm.py [hunk 5]
```diff
@@ -0,0 +1,1523 @@
...
+                    "function": {
+                        "name": part.function_call.name,
+                        "arguments": (
+                            json.dumps(part.function_call.args)
+                            if part.function_call.args
+                            else "{}"
+                        ),
+                    },
+                }
+            )
```
[the remark this answers is not in the document]
Comment (secprog) [reply src/google/adk_community/models/openai_llm.py]: fixed

File: src/google/adk_community/models/openai_llm.py [hunk 6]
```diff
@@ -0,0 +1,1523 @@
...
+        return "user"
+
+
+def _is_image_part(part: types.Part) -> bool:
+    """Checks if a part contains image data."""
+    return (
+        part.inline_data
+        and part.inline_data.mime_type
+        and part.inline_data.mime_type.startswith("image")
+    )
```
[the remark this answers is not in the document]
Comment (secprog) [reply src/google/adk_community/models/openai_llm.py]: fix

File: src/google/adk_community/models/openai_llm.py [hunk 7]
```diff
@@ -0,0 +1,1523 @@
...
+                            try:
+                                # Get schema name from the class/object
+                                if inspect.isclass(response_schema):
+                                    schema_name = getattr(response_schema, "__name__", "ResponseSchema")
+                                else:
+                                    schema_name = getattr(
+                                        response_schema, "__class__", type(response_schema)
+                                    ).__name__
+                                    if schema_name == "dict":
+                                        schema_name = "ResponseSchema"
```
[the remark this answers is not in the document]
Comment (secprog) [reply src/google/adk_community/models/openai_llm.py]: fixed

**Comment by DeanChensj** [comment]
@gemini-cli /review
